### PR TITLE
🔥Feature: Add support for TrustProxy

### LINKED ISSUE #2930: 📝 [Proposal]: Add Helpers for Internal IP Ranges
### Feature Proposal Description

Currently, adding loopback, link-local, and private network addresses to the `fiber.Config.TrustedProxies` list requires us to manually add those ranges. Since web apps are commonly deployed behind reverse proxies, it would be helpful to have a simple way of adding these ranges without needing to search them up.

When configuring trusted proxies in Echo framework, the setup is a bit different:
```go
e := echo.New()
_, myProxyRange, _ := net.ParseCIDR("173.245.48.0/20")
e.IPExtractor = echo.ExtractIPFromXFFHeader(
	echo.TrustLoopback(true), // e.g. ipv4 start with 127. 
	echo.TrustLinkLocal(true), // e.g. ipv4 start with 169.254
	echo.TrustPrivateNet(true), // e.g. ipv4 start with 10. or 192.168
	echo.TrustIPRange(myProxyRange),
)
```

In Fiber, an equivalent setup would look something like
```go
app := fiber.New(fiber.Config{
	ProxyHeader:             fiber.HeaderXForwardedFor,
	EnableTrustedProxyCheck: true,
	TrustedProxies: []string{
		"127.0.0.0/8",    // Loopback addresses
		"169.254.0.0/16", // Link-Local addresses
		"fe80::/10",
		"192.168.0.0/16", // Private Network addresses
		"172.16.0.0/12",
		"10.0.0.0/8",
		"fc00::/7",
		"173.245.48.0/20", // My custom range
	},
})
```

Although I prefer Fiber's method of using an array of strings, it would be nice to have constants like:
```go
// Fiber helpers.go
// Note: This is not an exhaustive list.
const (
        ...
	IPv4Loopback      = "127.0.0.0/8"
	IPv4LinkLocal     = "169.254.0.0/16"
	IPv4PrivateSmall  = "192.168.0.0/16"
	IPv4PrivateMedium = "172.16.0.0/12"
	IPv4PrivateLarge  = "10.0.0.0/8"
	IPv6Loopback      = "::1/128"
	IPv6LinkLocal     = "fe80::/10"
	IPv6PrivateNet    = "fc00::/7"
)
```

This would allow developers to use
```go
app := fiber.New(fiber.Config{
	ProxyHeader:             fiber.HeaderXForwardedFor,
	EnableTrustedProxyCheck: true,
	TrustedProxies: []string{
		fiber.IPv4Loopback,
		fiber.IPv4LinkLocal,
		fiber.IPv4PrivateSmall,
		fiber.IPv4PrivateMedium,
		fiber.IPv4PrivateLarge,
		fiber.IPv6Loopback,
		fiber.IPv6LinkLocal,
		fiber.IPv6PrivateNet,
		"173.245.48.0/20",  // My custom range
	},
})
```

They don't necessarily need to be constants. We could add a new config option to `fiber.Config` instead:
```go
app := fiber.New(fiber.Config{
	ProxyHeader:             fiber.HeaderXForwardedFor,
	EnableTrustedProxyCheck: true,
	TrustInternalIPs: true, // default to false
	TrustedProxies: []string{
		"173.245.48.0/20",  // My custom range
	},
})
```
This would be far easier, but at the cost of allowing developers to cherry pick individual ranges. Those cases are probably rare though, which might be why Echo's helper functions for these ranges [default to true](https://github.com/labstack/echo/blob/d549290448fc65ae075f0b960b25489f6a38cb78/ip.go#L154).

### Alignment with Express API

Express.js allows developers to [set trusted proxy settings using pre-defined subnets](https://expressjs.com/en/5x/api.html#trust.proxy.options.table):
```js
// loopback - 127.0.0.1/8, ::1/128
// linklocal - 169.254.0.0/16, fe80::/10
// uniquelocal - 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, fc00::/7
app.set('trust proxy', ['loopback', 'linklocal', 'uniquelocal']) 
```

### HTTP RFC Standards Compliance

n/a (this is a quality of life improvement)

### API Stability

IP address ranges don't change.

### Feature Examples

```go
(Listed above)
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have searched for existing issues that describe my proposal before opening this one.
- [X] I understand that a proposal that does not meet these guidelines may be closed without explanation.

### PR DESCRIPTION
**Pull request by xEricL**
# Description

This renames `EnableTrustedProxyCheck` to `TrustProxy`. It also adds a new config option called `TrustProxyConfig`, allowing developers to trust loopback, private, or link-local IP addresses without adding the ranges to the whitelist.

If these ranges are added manually, it can negatively impact performance (see https://github.com/gofiber/fiber/pull/2933 benchmarks).

I chose these three ranges because these are the same ranges enabled by default when configuring an [IP extractor in Echo](https://github.com/labstack/echo/blob/d20a6257aa1263681e2ce1f020f8d801a8be9782/ip.go#L174).

- Fixes #2930
- Replaces #3137 

## Changes introduced

- [x] Benchmarks: Describe any performance benchmarks and improvements related to the changes.
- [x] Documentation Update: Detail the updates made to the documentation and links to the changed files.
- [x] Changelog/What's New: Include a summary of the additions for the upcoming release notes.
- [x] Migration Guide: If necessary, provide a guide or steps for users to migrate their existing code to accommodate these changes.
- [x] API Alignment with Express: Explain how the changes align with the Express API.
- [x] Examples: Provide examples demonstrating the new features or changes in action.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Enhancement (improvement to existing features and functionality)
- [x] Documentation update (changes to documentation)
- [x] Performance improvement (non-breaking change which improves efficiency)

## Checklist

- [x] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [x] Conducted a self-review of the code and provided comments for complex or critical parts.
- [x] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [x] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [x] Ensured that new and existing unit tests pass locally with the changes.
- [x] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [x] Aimed for optimal performance with minimal allocations in the new code.
- [x] Provided benchmarks for the new code to analyze and improve upon.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -330,29 +330,31 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
...
 
-	// Read EnableTrustedProxyCheck doc.
+	// Read TrustProxy doc.
 	//
-	// Default: []string
-	TrustedProxies     []string `json:"trusted_proxies"`
-	trustedProxiesMap  map[string]struct{}
-	trustedProxyRanges []*net.IPNet
+	// Default: DefaultTrustProxyConfig
+	TrustProxyConfig TrustProxyConfig `json:"trusted_proxy_config"`
```
Comment (gaby) [inline app.go:357]: Tag has wrong name

File: ctx.go [hunk 2]
```diff
@@ -1819,20 +1819,26 @@ func (c *DefaultCtx) configDependentPaths() {
...
-	if !c.app.config.EnableTrustedProxyCheck {
+	if !c.app.config.TrustProxy {
 		return true
 	}
 
 	ip := c.fasthttp.RemoteIP()
 
-	if _, trusted := c.app.config.trustedProxiesMap[ip.String()]; trusted {
+	if (c.app.config.TrustProxyConfig.Private && ip.IsPrivate()) ||
+		(c.app.config.TrustProxyConfig.Loopback && ip.IsLoopback()) ||
```
Comment (gaby) [inline ctx.go:1832]: Move loopback to the beginning of the IF statement. That way it short-circuits. It's very likely most traffic will be proxied from loopback so.
1. Loopback
2. Private
3. LinkLocal

File: app.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in 7cb9bb2d1cc1222f38421e8a566f99ada2d0a211

File: app.go [hunk 3]
```diff
@@ -330,29 +330,31 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
...
 	//   3. c.Host() and c.Hostname() WON'T get value from X-Forwarded-Host header, fasthttp.Request.URI().Host()
 	//    will be used to get the hostname.
 	//
+	// To automatically trust all loopback, link-local, or private IP addresses,
+	// without manually adding them to the TrustProxyConfig.Proxies whitelist,
+	// you can set TrustProxyConfig.Loopback, TrustProxyConfig.LinkLocal, or TrustProxyConfig.Private to true.
+	//
 	// Default: false
-	EnableTrustedProxyCheck bool `json:"enable_trusted_proxy_check"`
+	TrustProxy bool `json:"enable_trusted_proxy_check"`
```
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in 7cb9bb2d1cc1222f38421e8a566f99ada2d0a211

File: app.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in 7cb9bb2d1cc1222f38421e8a566f99ada2d0a211

Comment (xEricL) [reply app.go]: fixed in 7cb9bb2d1cc1222f38421e8a566f99ada2d0a211

**Comment by gaby** [comment]
@coderabbitai full review

File: ctx_test.go [hunk 4]
```diff
@@ -1653,78 +1653,104 @@
...
 	{
 		app := New(Config{
-			EnableTrustedProxyCheck: true,
-
-			TrustedProxies: []string{
-				"0.0.0.1/31junk",
+			TrustProxy: true,
+			TrustProxyConfig: TrustProxyConfig{
+				Proxies: []string{"0.0.0.1/31junk"},
+			},
```
[the remark this answers is not in the document]
Comment (gaby) [reply ctx_test.go:1724]: This is a unit-test

File: app.go [hunk 5]
```diff
@@ -330,29 +330,31 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
...
 	// all headers that could be spoofed.
-	// If request ip in TrustedProxies whitelist then:
+	// If request ip in TrustProxyConfig.Proxies whitelist, then:
 	//   1. c.Scheme() get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header
 	//   2. c.IP() get value from ProxyHeader header.
 	//   3. c.Host() and c.Hostname() get value from X-Forwarded-Host header
-	// But if request ip NOT in Trusted Proxies whitelist then:
-	//   1. c.Scheme() WON't get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header,
+	// But if request ip NOT in TrustProxyConfig.Proxies whitelist, then:
+	//   1. c.Scheme() WON'T get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header,
```
[the remark this answers is not in the document]
Comment (gaby) [reply app.go:341]: @xEricL Can you add the suggestion above. Thanks

File: app.go [hunk 6]
```diff
@@ -330,29 +330,31 @@
 	// For example, the Host HTTP header is usually used to return the requested host.
 	// But when you’re behind a proxy, the actual host may be stored in an X-Forwarded-Host header.
 	//
-	// If you are behind a proxy, you should enable TrustedProxyCheck to prevent header spoofing.
-	// If you enable EnableTrustedProxyCheck and leave TrustedProxies empty Fiber will skip
+	// If you are behind a proxy, you should enable TrustProxy to prevent header spoofing.
+	// If you enable TrustProxy and do not provide a TrustProxyConfig, Fiber will skip
 	// all headers that could be spoofed.
-	// If request ip in TrustedProxies whitelist then:
+	// If request ip in TrustProxyConfig.Proxies whitelist, then:
```
[the remark this answers is not in the document]
Comment (gaby) [reply app.go:336]: Same here

File: app.go [hunk 7]
```diff
@@ -529,17 +561,17 @@ func New(config ...Config) *App {
...
 		_, ipNet, err := net.ParseCIDR(ipAddress)
 		if err != nil {
 			log.Warnf("IP range %q could not be parsed: %v", ipAddress, err)
 		} else {
-			app.config.trustedProxyRanges = append(app.config.trustedProxyRanges, ipNet)
+			app.config.TrustProxyConfig.ranges = append(app.config.TrustProxyConfig.ranges, ipNet)
 		}
 	} else {
-		app.config.trustedProxiesMap[ipAddress] = struct{}{}
+		app.config.TrustProxyConfig.ips[ipAddress] = struct{}{}
```
[the remark this answers is not in the document]
Comment (gaby) [reply app.go:574]: I believe we need to validate here too?

File: ctx.go [hunk 8]
```diff
@@ -1819,20 +1819,26 @@ func (c *DefaultCtx) configDependentPaths() {
 }
 
 // IsProxyTrusted checks trustworthiness of remote ip.
-// If EnableTrustedProxyCheck false, it returns true
+// If Config.TrustProxy false, it returns true
 // IsProxyTrusted can check remote ip by proxy ranges and ip map.
 func (c *DefaultCtx) IsProxyTrusted() bool {
-	if !c.app.config.EnableTrustedProxyCheck {
+	if !c.app.config.TrustProxy {
 		return true
```
[the remark this answers is not in the document]
Comment (gaby) [reply ctx.go:1826]: I agree this is confusing, but it's the way it used to be in v2.

File: app.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in b1bde8810025a5e4deca8682f0243e744f9d09c8

File: app.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in 9c12e4057942c43108df5114ae7403bc75abeb86

File: app.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go]: fixed in a7d8479bde968e887e6654cbaa0b25bfed361b9e

File: app.go [hunk 9]
```diff
@@ -385,6 +387,36 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
...
+	// Loopback enables trusting all loopback IP ranges (e.g., 127.0.0.0/8, ::1/128).
+	//
+	// Default: false
+	Loopback bool `json:"loopback"`
+
+	// Private enables trusting all private IP ranges (e.g., 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, fc00::/7).
+	//
+	// Default: false
+	Private bool `json:"private"`
+}
```
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go:418]: The fields `ips` and `ranges` are not meant to be exported.  @gaby can you confirm?

File: app.go [hunk 10]
```diff
@@ -330,29 +330,31 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
 	// For example, the Host HTTP header is usually used to return the requested host.
 	// But when you’re behind a proxy, the actual host may be stored in an X-Forwarded-Host header.
 	//
-	// If you are behind a proxy, you should enable TrustedProxyCheck to prevent header spoofing.
-	// If you enable EnableTrustedProxyCheck and leave TrustedProxies empty Fiber will skip
+	// If you are behind a proxy, you should enable TrustProxy to prevent header spoofing.
+	// If you enable TrustProxy and do not provide a TrustProxyConfig, Fiber will skip
```
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go:334]: the sentence continues on the next line.

File: app.go [hunk 11]
```diff
@@ -330,29 +330,31 @@ type Config struct { //nolint:govet // Aligning the struct fields is not necessa
...
 	// all headers that could be spoofed.
-	// If request ip in TrustedProxies whitelist then:
+	// If the request IP is in the TrustProxyConfig.Proxies allowlist, then:
 	//   1. c.Scheme() get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header
 	//   2. c.IP() get value from ProxyHeader header.
 	//   3. c.Host() and c.Hostname() get value from X-Forwarded-Host header
-	// But if request ip NOT in Trusted Proxies whitelist then:
-	//   1. c.Scheme() WON't get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header,
+	// But if the request IP is NOT in the TrustProxyConfig.Proxies allowlist, then:
+	//   1. c.Scheme() WON'T get value from X-Forwarded-Proto, X-Forwarded-Protocol, X-Forwarded-Ssl or X-Url-Scheme header,
```
[the remark this answers is not in the document]
Comment (xEricL) [reply app.go:341]: fixed in ec8292698eb0b2eaf0a37f1d36c7799203eeeb5c

**Review by gaby — APPROVED** [review]
LGTM

**Review by ReneWerner87 — APPROVED** [review]

**Comment by gaby** [comment]
@xEricL Thank you! 💪

**Pull request merged**